### PR TITLE
Feature: Get RTCPeerConnection stats

### PR DESCRIPTION
**Pull request by R-Delfino95**
New method in PeerConnection module which let to get the statistics related to the RTCPeerConnection.

### CODE REVIEW

File: packages/millicast-sdk/src/PeerConnectionStats.js [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    const statsObject = {
+      audio: {
+        inbound: {},
+        outbound: {}
+      },
+      video: {
+        inbound: {},
+        outbound: {}
+      }
+    }
```
Comment (emilsas) [inline packages/millicast-sdk/src/PeerConnectionStats.js:95]:
```suggestion
  parseStats (rawStats) {
    this.previousStats = this.stats
    const statsObject = {
      audio: {
        inbound: {},
        outbound: {}
      },
      video: {
        inbound: {},
        outbound: {}
      },
      raw: rawStats
    }
```

File: packages/millicast-sdk/src/PeerConnectionStats.js [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+          }
+          break
+        }
+        default:
+          break
+      }
+    }
+
+    this.previousStats = statsObject
+    return { raw: latestStats, ...statsObject }
```
Comment (emilsas) [inline packages/millicast-sdk/src/PeerConnectionStats.js:142]:
```suggestion
    this.stats = statsObject
```

File: packages/millicast-sdk/src/PeerConnectionStats.js [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+      const error = `Invalid interval value ${interval}`
+      logger.error(error)
+      throw new Error(error)
+    }
+    this.interval = intervalCast * 1000
+    this.timerInterval = setInterval(async () => {
+      logger.debug('New internal interval executed')
+      const stats = await this.peer.getStats()
+      this.stats = this.parseStats(stats)
+      this.firstIntervalExecuted = true
```
Comment (emilsas) [inline packages/millicast-sdk/src/PeerConnectionStats.js:63]:
```suggestion
      this.parseStats(stats)
```

File: packages/millicast-sdk/src/PeerConnectionStats.js [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    }, 1000)
+
+    this.emitInterval = setInterval(() => {
+      if (this.firstIntervalExecuted) {
```
Comment (emilsas) [inline packages/millicast-sdk/src/PeerConnectionStats.js:67]: Remove this flag `this.firstIntervalExecuted`

File: packages/millicast-sdk/src/PeerConnectionStats.js [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+          // Safari is missing mediaType and kind for 'inbound-rtp'
+          if (!['audio', 'video'].includes(mediaType)) {
+            if (report.id.includes('Video')) mediaType = 'video'
+            else mediaType = 'audio'
+          }
+          const additionalData = this.getBaseReportData(report, mediaType)
+
+          const previousBytes = this.previousStats ? this.previousStats[mediaType].inbound.bytesReceived : 0
+          additionalData.bitrate = 8 * (report.bytesReceived - previousBytes)
```
Comment (emilsas) [inline packages/millicast-sdk/src/PeerConnectionStats.js:126]: Set bitrate 0 if we didn't have previous Stats

File: packages/millicast-sdk/src/PeerConnectionStats.js [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+    for (const report of latestStats.values()) {
+      switch (report.type) {
+        case 'outbound-rtp': {
+          const mediaType = this.getMediaType(report)
+          const codecInfo = this.getCodecData(report.codecId, latestStats)
+          const additionalData = this.getBaseReportData(report, mediaType)
+
+          const previousBytes = this.previousStats ? this.previousStats[mediaType].outbound.bytesSent : 0
+          additionalData.bitrate = 8 * (report.bytesSent - previousBytes)
```
Comment (emilsas) [inline packages/millicast-sdk/src/PeerConnectionStats.js:105]: Set bitrate 0 if we didn't have previous Stats

File: packages/millicast-sdk/src/PeerConnectionStats.js [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+      codecData.mimeType = codecReport.mimeType
+    }
+    return codecData
+  }
+
+  /**
+   * Get common information
+   * @param {Object} report - JSON object which represents a report from RTCPeerConnection stats.
+   * @param {String} mediaType - Media type.
+   * @returns Object containing common information.
```
Comment (emilsas) [inline packages/millicast-sdk/src/PeerConnectionStats.js:173]:
```suggestion
   * @returns {Object} Object containing common information.
```

File: packages/millicast-sdk/src/PeerConnectionStats.js [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    this.firstIntervalExecuted = false
+    this.timerInterval = null
+    this.emitInterval = null
+    this.previousStats = null
+    this.interval = null
+  }
+
+  /**
+   * Initialize the statistics monitoring of the RTCPeerConnection.
+   * @param {Number|String} interval - Interval in seconds of how often it should get stats.
```
Comment (emilsas) [inline packages/millicast-sdk/src/PeerConnectionStats.js:48]:
```suggestion
   * @param {Number} interval - Interval in seconds of how often it should get stats.
```

File: packages/millicast-sdk/src/PeerConnection.js [hunk 9]
```diff
@@ -318,6 +322,61 @@ export default class PeerConnection extends EventEmitter {
   getTracks () {
     return this.peer?.getSenders()?.map((sender) => sender.track)
   }
+
+  /**
+   * Initialize the statistics monitoring of the RTCPeerConnection.
+   * @param {Number|String} interval - Interval in seconds of how often it should get stats.
```
Comment (emilsas) [inline packages/millicast-sdk/src/PeerConnection.js:328]:
```suggestion
   * @param {Number} interval - Interval in seconds of how often it should get stats. Should be greater or equals to 1 second.
```

File: packages/millicast-sdk/src/PeerConnection.js [hunk 10]
```diff
@@ -318,6 +322,61 @@ export default class PeerConnection extends EventEmitter {
...
+   *
+   * //Initialize the get stats with 2 seconds interval
+   * millicastView.webRTCPeer.getStats(2)
+   *
+   * //Capture new stats from event every 2 seconds
+   * millicastView.webRTCPeer.on('stats', (stats) => {
+   *   console.log('Stats from event: ', stats)
+   * })
+   */
+  getStats (interval = 1) {
```
Comment (emilsas) [inline packages/millicast-sdk/src/PeerConnection.js:360]:
```suggestion
  initStats (interval) {
```

File: packages/millicast-sdk/src/PeerConnectionStats.js [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+    return report.mediaType || report.kind
+  }
+
+  /**
+   * Get codec information from stats.
+   * @param {String} codecReportId - Codec report ID.
+   * @param {RTCStatsReport} latestStats - RTCPeerConnection stats.
+   * @returns {Object} Object containing codec information.
+   */
+  getCodecData (codecReportId, latestStats) {
```
Comment (emilsas) [inline packages/millicast-sdk/src/PeerConnectionStats.js:160]: Capture mimetype in codecReport
`const { mimeType } =  codecReportId ? latestStats.get(codecReportId) : {}`

**Review by emilsas — CHANGES_REQUESTED** [review]
In PeerConnectionStats, refactor all methods that don't use `this` to static.

**Comment by R-Delfino95** [comment]
@emilsas your suggestions were applied.

File: packages/millicast-sdk/src/PeerConnection.js [hunk 12]
```diff
@@ -318,6 +322,62 @@ export default class PeerConnection extends EventEmitter {
...
+   * millicastView.webRTCPeer.initStats(2)
+   *
+   * //Capture new stats from event every 2 seconds
+   * millicastView.webRTCPeer.on('stats', (stats) => {
+   *   console.log('Stats from event: ', stats)
+   * })
+   */
+  initStats (interval) {
+    if (this.peerConnectionStats) {
+      logger.warn('Cannot get peer stats: Already initialized')
```
Comment (emilsas) [inline packages/millicast-sdk/src/PeerConnection.js:363]:
```suggestion
      logger.warn('Cannot init peer stats: Already initialized')
```

File: packages/millicast-sdk/src/PeerConnection.js [hunk 13]
```diff
@@ -318,6 +322,62 @@ export default class PeerConnection extends EventEmitter {
...
+    } else if (this.peer) {
+      this.peerConnectionStats = new PeerConnectionStats(this.peer)
+      this.peerConnectionStats.init(interval)
+      reemit(this.peerConnectionStats, this, [peerConnectionStatsEvents.stats])
+    } else {
+      logger.warn('Cannot get peer stats: RTCPeerConnection not initialized')
```
Comment (emilsas) [inline packages/millicast-sdk/src/PeerConnection.js:369]:
```suggestion
      logger.warn('Cannot init peer stats: RTCPeerConnection not initialized')
```

**Review by emilsas — CHANGES_REQUESTED** [review]

**Review by emilsas — APPROVED** [review]

**Pull request merged**